File: src/bus_sim/oracle/actions.py
"""The fixed 221-slot action table defined by schema v2."""

from __future__ import annotations

from bus_sim.oracle.domain import Action

F_MAX, R_MAX = 16, 4
HEADWAYS = (360, 600, 900)


def build_action_table() -> tuple[Action, ...]:
    actions = [Action()]
    actions += [Action("DISPATCH", b, r) for b in range(F_MAX) for r in range(R_MAX)]
    actions += [Action("REASSIGN", b, r) for b in range(F_MAX) for r in range(R_MAX)]
    actions += [Action("SHORT_TURN", b, r) for b in range(F_MAX) for r in range(R_MAX)]
    actions += [Action("RECALL", b) for b in range(F_MAX)]
    actions += [
        Action("SET_HEADWAY", route_id=r, headway_s=h) for r in range(R_MAX) for h in HEADWAYS
    ]
    assert len(actions) == 221
    return tuple(actions)


ACTION_TABLE = build_action_table()


def action_id(
    kind: str, bus_id: int | None = None, route_id: int | None = None, headway_s: int | None = None
) -> int:
    wanted = Action(kind, bus_id, route_id, headway_s)
    try:
        return ACTION_TABLE.index(wanted)
    except ValueError as error:
        raise ValueError(f"unknown action: {wanted}") from error

File: crates/bus-sim-core/src/actions.rs
//! The fixed 221-slot action table (`schema v2`), ported from
//! `src/bus_sim/oracle/actions.py`.

pub const F_MAX: usize = 16;
pub const R_MAX: usize = 4;
pub const HEADWAYS: [i64; 3] = [360, 600, 900];
pub const ACTION_COUNT: usize = 221;

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum Action {
    Noop,
    Dispatch { bus: usize, route: u32 },
    Reassign { bus: usize, route: u32 },
    ShortTurn { bus: usize, route: u32 },
    Recall { bus: usize },
    SetHeadway { route: u32, headway_s: i64 },
}

impl Action {
    pub fn kind(self) -> &'static str {
        match self {
            Action::Noop => "NOOP",
            Action::Dispatch { .. } => "DISPATCH",
            Action::Reassign { .. } => "REASSIGN",
            Action::ShortTurn { .. } => "SHORT_TURN",
            Action::Recall { .. } => "RECALL",
            Action::SetHeadway { .. } => "SET_HEADWAY",
        }
    }
}

/// Decode an action index into the frozen table layout.
pub fn action_at(index: usize) -> Option<Action> {
    if index == 0 {
        return Some(Action::Noop);
    }
    let mut cursor = index - 1;
    if cursor < F_MAX * R_MAX {
        return Some(Action::Dispatch {
            bus: cursor / R_MAX,
            route: (cursor % R_MAX) as u32,
        });
    }
    cursor -= F_MAX * R_MAX;
    if cursor < F_MAX * R_MAX {
        return Some(Action::Reassign {
            bus: cursor / R_MAX,
            route: (cursor % R_MAX) as u32,
        });
    }
    cursor -= F_MAX * R_MAX;
    if cursor < F_MAX * R_MAX {
        return Some(Action::ShortTurn {
            bus: cursor / R_MAX,
            route: (cursor % R_MAX) as u32,
        });
    }
    cursor -= F_MAX * R_MAX;
    if cursor < F_MAX {
        return Some(Action::Recall { bus: cursor });
    }
    cursor -= F_MAX;
    if cursor < R_MAX * HEADWAYS.len() {
        return Some(Action::SetHeadway {
            route: (cursor / HEADWAYS.len()) as u32,
            headway_s: HEADWAYS[cursor % HEADWAYS.len()],
        });
    }
    None
}
